r/R/dplyr-funcs-datetime.R

```diff
@@ -147,5 +147,16 @@ register_bindings_datetime <- function() {
   register_binding("pm", function(x) {
     !call_binding("am", x)
   })
+  register_binding("tz", function(x) {
+    if (!call_binding("is.POSIXct", x)) {
+      if (inherits(x, "Expression")) {
+        class <- x$type()$ToString()
+      } else {
+        class <- type(x)
+      }
+      abort(paste0("timezone extraction for objects of class `", class, "` not supported in Arrow"))
+    }
```

Comment:
I had to make a few changes here to make sure that this works with both R objects and Arrow objects. I only discovered this in my own testing. We probably won't see many pure R objects here, but it's possible!

We don't (AFAIK) have a unified way to get types that would work from R objects and Arrow objects, but it might be worth looking for one + making a jira if we don't have one already.

Reply:
Oh also I shuffled around the `if` clause so that we error earlier in the body and then have a standard return if `x` gets through all the validation. This is functionally the same as what we had before, but I find matches the style of others a bit more (and IMO if we can avoid an early return in a function, it's good to)